### PR TITLE
openfaas output type added

### PR DESCRIPTION
**Pull request by developer-guy**
<Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area config

/area outputs

> /area tests

**What this PR does / why we need it**:

This PR adds OpenFaaS as an output module. So, we can start sending events to the OpenFaaS functions. 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by developer-guy** [comment]
cc @dentrax

**Review by cpanato — CHANGES_REQUESTED** [review]
thanks fora dding that, in general lgtm, but can you add the the new output in the readme and in the config_example as well?

**Comment by developer-guy** [comment]
> thanks fora dding that, in general lgtm, but can you add the the new output in the readme and in the config_example as well?

Of course @cpanato, it is not completed yet, I'm going to complete it tonight 🙋🏻‍♂️

**Comment by developer-guy** [comment]
@cpanato, the PR is ready to review now 👍

File: config.go [hunk 1]
```diff
@@ -141,6 +141,15 @@ func getConfig() *types.Configuration {
 	v.SetDefault("Kubeless.Port", 8080)
 	v.SetDefault("Kubeless.Kubeconfig", "")
 	v.SetDefault("Kubeless.MinimumPriority", "")
+
+	v.SetDefault("Openfaas.GatewayNamespace", "openfaas")
+	v.SetDefault("Openfaas.GatewayAddress", "gateway")
```
Comment (Issif) [inline config.go:146]:
```suggestion
	v.SetDefault("Openfaas.GatewayService", "gateway")
```
or

```suggestion
	v.SetDefault("Openfaas.GatewayName", "gateway")
```

I think the term `Address` is confusing, we expect something like `http://xxxx` when we see that pattern. In our case, it's a kubernetes service so we only need its name.

My favorite is `GatewayService`, I find it more explicit.

File: config.go [hunk 2]
```diff
@@ -141,6 +141,15 @@ func getConfig() *types.Configuration {
...
+	v.SetDefault("Openfaas.FunctionName", "")
+	v.SetDefault("Openfaas.FunctionNamespace", "openfaas-fn")
+	v.SetDefault("Openfaas.Port", 8080)
```
Comment (Issif) [inline config.go:149]:
```suggestion
	v.SetDefault("Openfaas.GatewayPort", 8080)
```

It makes more sense as it's related to Gateway

File: config.go (hunk 1 above)
Comment (developer-guy) [reply config.go:146]: yep, GatewayService would be a better name for  the field 👍

File: config.go (hunk 2 above)
Comment (developer-guy) [reply config.go:149]: Yep, you are right 👍

File: config.go (hunk 1 above)
Comment (developer-guy) [reply config.go]: done.

File: config.go (hunk 2 above)
Comment (developer-guy) [reply config.go]: done.

File: config_example.yaml [hunk 3]
```diff
@@ -166,5 +166,14 @@ kubeless:
   kubeconfig: "~/.kube/config" # Kubeconfig file to use (only if falcoside is running outside the cluster)
   # minimumpriority: "debug" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
 
+openfaas:
+  gatewaynamespace: "" # Namespace of OpenFaaS Gateway, "openfaas" (default)
+  gatewayservice: "" # Service of OpenFaaS Gateway, "gateway" (default)
+  functionname: "" # Name of OpenFaaS function, if not empty, OpenFaaS is enabled
+  functionnamespace: "" # Namespace of OpenFaaS function, "openfaas-fn" (default)
+  gatewayport: 8080 # Port of service of OpenFaaS Gateway
+  kubeconfig: "~/.kube/config" # Kubeconfig file to use (only if falcosidekick is running outside the cluster)
```
Comment (Issif) [inline config_example.yaml:175]:
```suggestion
  gatewayservice: "" # Service of OpenFaaS Gateway, "gateway" (default)
  gatewaynamespace: "" # Namespace of OpenFaaS Gateway, "openfaas" (default)
  gatewayport: 8080 # Port of service of OpenFaaS Gateway
  functionname: "" # Name of OpenFaaS function, if not empty, OpenFaaS is enabled
  functionnamespace: "" # Namespace of OpenFaaS function, "openfaas-fn" (default)
  kubeconfig: "~/.kube/config" # Kubeconfig file to use (only if falcosidekick is running outside the cluster)
```

better order

File: outputs/openfaas.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		str, _ := json.Marshal(falcopayload)
+		req := c.KubernetesClient.CoreV1().RESTClient().Post().AbsPath("/api/v1/namespaces/" + c.Config.Openfaas.GatewayNamespace + "/services/" + c.Config.Openfaas.GatewayService + ":" + strconv.Itoa(c.Config.Openfaas.GatewayPort) + "/proxy" + "/function/" + c.Config.Openfaas.FunctionName + "." + c.Config.Openfaas.FunctionNamespace).Body(str)
+		req.SetHeader("event-id", uuid.New().String())
+		req.SetHeader("Content-Type", "application/json")
+		req.SetHeader("User-Agent", "Falcosidekick")
+
+		res := req.Do(context.TODO())
+		rawbody, err := res.Raw()
+		if err != nil {
+			go c.CountMetric(Outputs, 1, []string{"output:OpenFaaS", "status:error"})
```
Comment (Issif) [inline outputs/openfaas.go:63]:
```suggestion
			go c.CountMetric(Outputs, 1, []string{"output:openfaaS", "status:error"})
```

File: outputs/openfaas.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+			c.Stats.Openfaas.Add(Error, 1)
+			c.PromStats.Outputs.With(map[string]string{"destination": "OpenFaaS", "status": Error}).Inc()
```
Comment (Issif) [inline outputs/openfaas.go:65]:
```suggestion
			c.PromStats.Outputs.With(map[string]string{"destination": "openfaaS", "status": Error}).Inc()
```

File: outputs/openfaas.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+			log.Printf("[ERROR] : Openfaas - %v\n", err)
```
Comment (Issif) [inline outputs/openfaas.go:66]:
```suggestion
			log.Printf("[ERROR] : OpenFaas - %v\n", err)
```

File: outputs/openfaas.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+			return
+		}
+		log.Printf("[INFO]  : OpenFaaS - Function Response : %v\n", string(rawbody))
+	} else {
+		err := c.Post(falcopayload)
+		if err != nil {
+			go c.CountMetric(Outputs, 1, []string{"output:OpenFaaS", "status:error"})
```
Comment (Issif) [inline outputs/openfaas.go:73]:
```suggestion
			go c.CountMetric(Outputs, 1, []string{"output:openfaas", "status:error"})
```

File: outputs/openfaas.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+			c.Stats.Openfaas.Add(Error, 1)
+			c.PromStats.Outputs.With(map[string]string{"destination": "OpenFaaS", "status": Error}).Inc()
```
Comment (Issif) [inline outputs/openfaas.go:75]:
```suggestion
			c.PromStats.Outputs.With(map[string]string{"destination": "openfaas", "status": Error}).Inc()
```

File: outputs/openfaas.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+			log.Printf("[ERROR] : OpenFaaS - %v\n", err)
+			return
+		}
+	}
+	log.Printf("[INFO]  : OpenFaaS - Call Function \"%v\" OK\n", c.Config.Openfaas.FunctionName+"."+c.Config.Openfaas.FunctionNamespace)
+	go c.CountMetric(Outputs, 1, []string{"output:OpenFaaS", "status:ok"})
```
Comment (Issif) [inline outputs/openfaas.go:81]:
```suggestion
	go c.CountMetric(Outputs, 1, []string{"output:openfaas", "status:ok"})
```

File: outputs/openfaas.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	c.Stats.Openfaas.Add(OK, 1)
+	c.PromStats.Outputs.With(map[string]string{"destination": "OpenFaaS", "status": OK}).Inc()
```
Comment (Issif) [inline outputs/openfaas.go:83]:
```suggestion
	c.PromStats.Outputs.With(map[string]string{"destination": "openfaas", "status": OK}).Inc()
```

File: outputs/constants.go [hunk 11]
```diff
@@ -37,4 +37,5 @@ const (
 	Orange    string = "#ff5400"
 
 	Kubeless string = "Kubeless"
+	Openfaas string = "Openfaas"
```
Comment (Issif) [inline outputs/constants.go:40]: I didn't found out where you use this constant.

For `Kubeless` it's used [HERE](https://github.com/falcosecurity/falcosidekick/blob/c3962df9b4e7d9f6943008c12cfe6a9193a548cc/outputs/client.go#L166)

I think the same code block is missing for your `OpenFaas` client, without, the response will not be logged.

**Review by Issif — CHANGES_REQUESTED** [review]
Can you update the README too, please.

* Add `OpenFaas` in list of available outputs
* Add settings in configuration
	* in extract of `config_example.yaml`
	* in table of env vars

File: outputs/constants.go (hunk 11 above)
Comment (developer-guy) [reply outputs/constants.go:40]: done, thank you, we missed it.

File: outputs/openfaas.go (hunk 10 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 9 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 8 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 7 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 6 above)
Comment (developer-guy) [reply outputs/openfaas.go:66]: done

File: outputs/openfaas.go (hunk 5 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 4 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: config_example.yaml (hunk 3 above)
Comment (developer-guy) [reply config_example.yaml]: nice, done.

File: outputs/client.go [hunk 12]
```diff
@@ -138,6 +138,12 @@ func (c *Client) Post(payload interface{}) error {
 		req.Header.Add("event-namespace", c.Config.Kubeless.Namespace)
 	}
 
+	if c.OutputType == Openfaas {
+		req.Header.Add("event-id", uuid.New().String())
+		req.Header.Add("event-type", "falco")
+		req.Header.Add("event-namespace", c.Config.Openfaas.FunctionNamespace)
+	}
+
```
Comment (Issif) [inline outputs/client.go:146]: I don't think these headers are mandatory for OpenFaas as they are for Kubeless. If you want to set them, maybe we could merge with Kubeless condition.

```go
	if c.OutputType == Kubeless || c.OutputType == Openfaas {
		req.Header.Add("event-id", uuid.New().String())
		req.Header.Add("event-type", "falco")
        switch c.OutputType {
		case Kubeless:
			req.Header.Add("event-namespace", c.Config.Kubeless.Namespace)
		case Openfass:
			req.Header.Add("event-namespace", c.Config.Openfaas.FunctionNamespace)	
		}
	}
```

Comment (developer-guy) [reply outputs/client.go:146]: yeah, these are not mandatory for OpenFaaS, let me clean this up 👍

**Review by Issif — CHANGES_REQUESTED** [review]
Still remains to complete:
- https://github.com/falcosecurity/falcosidekick/blob/master/README.md#yaml-file
- https://github.com/falcosecurity/falcosidekick/blob/master/README.md#env-vars

**Comment by developer-guy** [comment]
> Still remains to complete:
> 
> * https://github.com/falcosecurity/falcosidekick/blob/master/README.md#yaml-file
> * https://github.com/falcosecurity/falcosidekick/blob/master/README.md#env-vars

Completed 👍

File: outputs/client.go [hunk 13]
```diff
@@ -162,6 +162,9 @@ func (c *Client) Post(payload interface{}) error {
 		if c.OutputType == Kubeless {
 			body, _ := ioutil.ReadAll(resp.Body)
 			log.Printf("[INFO]  : Kubeless - Function Response : %v\n", string(body))
+		} else if c.OutputType == Openfaas {
+			body, _ := ioutil.ReadAll(resp.Body)
```
Comment (Issif) [inline outputs/client.go:166]: you can move this line before the `if` for avoiding duplication of code

File: outputs/openfaas.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+			Config:           config,
+			Stats:            stats,
+			PromStats:        promStats,
+			StatsdClient:     statsdClient,
+			DogstatsdClient:  dogstatsdClient,
+			KubernetesClient: clientset,
+		}, nil
+	}
+	return NewClient(
+		"Openfaas",
```
Comment (Issif) [inline outputs/openfaas.go:39]:
```suggestion
		Openfaas,
```
same

File: outputs/openfaas.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+		restConfig, err := clientcmd.BuildConfigFromFlags("", config.Openfaas.Kubeconfig)
+		if err != nil {
+			return nil, err
+		}
+		clientset, err := kubernetes.NewForConfig(restConfig)
+		if err != nil {
+			return nil, err
+		}
+		return &Client{
+			OutputType:       "OpenFaaS",
```
Comment (Issif) [inline outputs/openfaas.go:29]:
```suggestion
			OutputType:       Openfaas,
```

We can use the constant

File: outputs/openfaas.go [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+		res := req.Do(context.TODO())
+		rawbody, err := res.Raw()
+		if err != nil {
+			go c.CountMetric(Outputs, 1, []string{"output:openfaas", "status:error"})
+			c.Stats.Openfaas.Add(Error, 1)
+			c.PromStats.Outputs.With(map[string]string{"destination": "openfaas", "status": Error}).Inc()
+			log.Printf("[ERROR] : Openfaas - %v\n", err)
+			return
+		}
+		log.Printf("[INFO]  : Openfaas - Function Response : %v\n", string(rawbody))
```
Comment (Issif) [inline outputs/openfaas.go:69]:
```suggestion
		log.Printf("[INFO]  : %v - Function Response : %v\n", Openfaas, string(rawbody))
```

File: outputs/openfaas.go [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+	} else {
+		err := c.Post(falcopayload)
+		if err != nil {
+			go c.CountMetric(Outputs, 1, []string{"output:openfaas", "status:error"})
+			c.Stats.Openfaas.Add(Error, 1)
+			c.PromStats.Outputs.With(map[string]string{"destination": "openfaas", "status": Error}).Inc()
+			log.Printf("[ERROR] : Openfaas - %v\n", err)
```
Comment (Issif) [inline outputs/openfaas.go:76]:
```suggestion
			log.Printf("[ERROR] : %v - %v\n", Openfaas, err)
```

File: outputs/openfaas.go [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+			return
+		}
+	}
+	log.Printf("[INFO]  : Openfaas - Call Function \"%v\" OK\n", c.Config.Openfaas.FunctionName+"."+c.Config.Openfaas.FunctionNamespace)
```
Comment (Issif) [inline outputs/openfaas.go:80]:
```suggestion
	log.Printf("[INFO]  : %v - Call Function \"%v\" OK\n", Openfaas, c.Config.Openfaas.FunctionName+"."+c.Config.Openfaas.FunctionNamespace)
```

File: outputs/openfaas.go (hunk 6 above)
Comment (Issif) [reply outputs/openfaas.go:66]:
```suggestion
			log.Printf("[ERROR] : %v - %v\n", Openfaas, err)
```

File: outputs/constants.go (hunk 11 above)
Comment (Issif) [inline outputs/constants.go:40]:
```suggestion
	Openfaas string = "OpenFaas"
```

File: outputs/client.go [hunk 19]
```diff
@@ -162,6 +162,9 @@ func (c *Client) Post(payload interface{}) error {
...
+			log.Printf("[INFO]  : Openfaas - Function Response : %v\n", string(body))
```
Comment (Issif) [inline outputs/client.go:167]:
```suggestion
			log.Printf("[INFO]  : %v - Function Response : %v\n", Openfaas, string(body))
```

**Review by Issif — CHANGES_REQUESTED** [review]

File: outputs/openfaas.go (hunk 14 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 16 above)
Comment (developer-guy) [reply outputs/openfaas.go:69]: done

File: outputs/openfaas.go (hunk 15 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/client.go (hunk 13 above)
Comment (developer-guy) [reply outputs/client.go]: done

File: outputs/openfaas.go (hunk 17 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 18 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/constants.go (hunk 11 above)
Comment (developer-guy) [reply outputs/constants.go]: done

File: outputs/client.go (hunk 19 above)
Comment (developer-guy) [reply outputs/client.go]: done

File: outputs/constants.go (hunk 11 above)
Comment (developer-guy) [reply outputs/constants.go]: done

File: outputs/openfaas.go (hunk 18 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 17 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 16 above)
Comment (developer-guy) [reply outputs/openfaas.go:69]: done

File: outputs/openfaas.go (hunk 15 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/openfaas.go (hunk 14 above)
Comment (developer-guy) [reply outputs/openfaas.go]: done

File: outputs/client.go (hunk 13 above)
Comment (developer-guy) [reply outputs/client.go]: done

**Review by Issif — APPROVED** [review]
🎉 

Thanks for this amazing work and all your patience with my huge amount of comments. Really good job, I'm glad you help us.

**Review by Issif — APPROVED** [review]

**Comment by developer-guy** [comment]
> 🎉 
> 
> 
> 
> Thanks for this amazing work and all your patience with my huge amount of comments. Really good job, I'm glad you help us.

It was a honor for me, thank you, I'm really appreciate it🙏🙏

**Review by cpanato — APPROVED** [review]
/lgtm

nice work!! thanks so much for that!

**Pull request merged**